developer/js/lexical-model-compiler/build-trie.ts

```diff
@@ -65,6 +75,58 @@ export function parseWordList(contents: string): WordList {
   return result;
 }
 
+export function compileWordListCharacterSet(wordlist: WordList, searchTermToKey?: (wf: string) => string): string {
+  let charMap: {[char: string]: number} = {};
+  let charList: string[] = [];
+
+  for(let [word, weight] of wordlist) {
+    // There's a problem if we have `null` or `undefined` for a word.
+    if(!word) {
+      throw TypeError();
+    }
+
+    // We want to operate on the search-term form, since that's what is used to look up the words in the Trie.
+    if(searchTermToKey) {
+      word = searchTermToKey(word);
+    }
+
+    for(let i=0; i < word.length; i++) {
+      let code = word.charCodeAt(i);
+      let char = word.charAt(i);
+
+      // Check for surrogate pairs
+      if(code >= 0xD800 && code <= 0xDBFF && word.length > i+1) {
+        let pairedCode = word.charCodeAt(i+1);
+
+        // Ensure it's actually a legit paired code
+        if(pairedCode >= 0xDC00 && pairedCode <= 0xDFFF) {
+          // Computes the actual character's code.
+          //code = (code - 0xD800) * 0x400 + pairedCode - 0xDC00 + 0x10000
+
+          // Only process the second code if it's actually paired.
+          char = char + word.charAt(++i); // The second code is processed, so skip it in the loop.
+        } // else handle pairedCode in the next loop iteration.
+      }
+      
+      if(!charMap[char]) {
+        charMap[char] = weight;
+        charList.push(char);
+      } else {
+        charMap[char] += weight;
+      }
+    }
+  }
+
+  charList.sort();
+
+  let charSet: CharacterSet = [];
+  for(let c of charList) {
+    charSet.push([c, charMap[c]]);
+  }
+
+  return JSON.stringify(charSet);
```

Comment:
I think it is more appropriate for this function to return the character set itself, rather than a JSON-ified structure. I think the code would be better factored if the stringification happened in the compiler code, rather than this otherwise reusable and testable method. The compiler routine is the one that actually cares about stringifying source code!

Reply:
What about `createTrieDataStructure`, then?  I was merely following the convention already established by that method.  It'd be a bit odd for the two to operate differently in this regard.

... I actually thought about adjusting said method according to your advice here and doing it the other way with `compileWordListCharacterSet` as well, but didn't want to add unnecessary changes.

Reply:
Now that we have testing infrastructure, I think **both** `createTrieDataStructure()` and `compileWordListCharacterSet()` should return JavaScript objects rather than strings.

We can save this for another PR, though.